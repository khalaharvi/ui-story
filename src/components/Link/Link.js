import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styled, { css, cx as emotionCx } from 'react-emotion';

import { createClassNames } from 'utils';

const cx = createClassNames('link');

const isSpecialClick = (event) => {
  const isMiddleClick = event.button === 1;
  return Boolean(
    isMiddleClick ||
      event.altKey ||
      event.ctrlKey ||
      event.metaKey ||
      event.shiftKey
  );
};

export const StyledLink = styled('a')({
  textDecoration: 'none'
});

const Link = ({
  className,
  onClick,
  to,
  children,
  external,
  noFollow
}) => {
  const clickHandler = (e) => {
    if (isSpecialClick(e)) {
      return;
    }
    onClick();
    e.preventDefault();
  };

  let rel = [];
  let target;

  if (external) {
    rel = ['external', 'noopener'];
    target = '_blank';
  }

  if (noFollow) {
    rel.push('nofollow');
  }

  const attrs = {
    rel: rel.length > 0 ? rel.join(' ') : undefined,
    target
  };

  return (
    <StyledLink
      href={to}
      className={classnames(cx(''), className)}
      onClick={onClick ? clickHandler : null}
      {...attrs}
    >
      {children}
    </StyledLink>
  );
};

Link.propTypes = {
  className: PropTypes.string,
  to: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.node
  ]).isRequired,
  onClick: PropTypes.func,
  external: PropTypes.bool,
  noFollow: PropTypes.bool
};

Link.defaultProps = {
  external: false,
  noFollow: false
};

const linkWithIcon = css({
  display: 'flex',
  alignItems: 'center',

  'span:first-child': {
    display: 'inline-block',
    width: '17px',
    marginRight: '12px'
  },

  svg: {
    fill: 'currentColor',
    display: 'block',
    margin: '0 auto'
  }
});

export const LinkWithIcon = ({ icon, label, ...props }) => (
  <Link className={emotionCx(cx('-icon'), linkWithIcon)} {...props}>
    <span>{icon}</span>
    <span>{label}</span>
  </Link>
);

LinkWithIcon.propTypes = {
  icon: PropTypes.element.isRequired,
  label: PropTypes.string.isRequired
};

export default Link;
