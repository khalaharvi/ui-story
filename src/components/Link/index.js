export {
  default,
  LinkWithIcon
} from './Link';
