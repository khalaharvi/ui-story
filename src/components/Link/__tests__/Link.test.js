import React from 'react';
import { mount, shallow } from 'enzyme';

import Link, { LinkWithIcon } from '../Link';

describe('<Link />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Link to="https://deringhall.com">Visit Dering Hall</Link>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with a child element', () => {
    const Element = () => (<span>Visit Dering Hall</span>);
    const wrapper = shallow(<Link to="https://deringhall.com"><Element /></Link>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with className prop', () => {
    const wrapper = shallow(<Link to="https://deringhall.com" className="custom-classname">Visit Dering Hall</Link>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render an <a> with an href attribute', () => {
    const wrapper = mount(<Link to="https://deringhall.com/">Visit Dering Hall</Link>);

    const a = wrapper.find('a');

    expect(a).toHaveLength(1);
    expect(wrapper.getDOMNode().href).toBe('https://deringhall.com/');

    wrapper.unmount();
  });

  it('should have a rel="external noopener" and target="_blank" attributes when external', () => {
    const wrapper = mount(<Link to="https://deringhall" external>Visit Dering Hall</Link>);

    const a = wrapper.find('a');

    expect(a).toHaveLength(1);
    expect(wrapper.getDOMNode().rel).toBe('external noopener');
    expect(wrapper.getDOMNode().target).toBe('_blank');

    wrapper.unmount();
  });

  it('should have a rel="nofollow" attribute when noFollow', () => {
    const wrapper = mount(<Link to="https://deringhall" noFollow>Visit Dering Hall</Link>);

    const a = wrapper.find('a');

    expect(a).toHaveLength(1);
    expect(wrapper.getDOMNode().rel).toBe('nofollow');

    wrapper.unmount();
  });

  it('should have a rel="external noopener nofollow" and target="_blank" attributes when external and noFollow', () => {
    const wrapper = mount(<Link to="https://deringhall" external noFollow>Visit Dering Hall</Link>);

    const a = wrapper.find('a');

    expect(a).toHaveLength(1);
    expect(wrapper.getDOMNode().rel).toBe('external noopener nofollow');
    expect(wrapper.getDOMNode().target).toBe('_blank');

    wrapper.unmount();
  });

  it('calls onClick when it has an onClick prop', () => {
    const onClick = jest.fn();
    const wrapper = mount(<Link to="https://deringhall.com/" onClick={onClick}>Visit Dering Hall</Link>);

    const a = wrapper.find('a');

    a.simulate('click');
    expect(onClick.mock.calls).toHaveLength(1);

    wrapper.unmount();
  });
});

describe('<LinkWithIcon />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(
      <LinkWithIcon
        to="https://deringhall.com"
        icon={<i>custom-icon</i>}
        label="Label"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
