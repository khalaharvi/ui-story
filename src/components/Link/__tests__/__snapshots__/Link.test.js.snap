// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> renders correctly 1`] = `
<Styled(a)
  className="dh-link"
  href="https://deringhall.com"
  onClick={null}
>
  Visit Dering Hall
</Styled(a)>
`;

exports[`<Link /> renders correctly with a child element 1`] = `
<Styled(a)
  className="dh-link"
  href="https://deringhall.com"
  onClick={null}
>
  <Element />
</Styled(a)>
`;

exports[`<Link /> renders correctly with className prop 1`] = `
<Styled(a)
  className="dh-link custom-classname"
  href="https://deringhall.com"
  onClick={null}
>
  Visit Dering Hall
</Styled(a)>
`;

exports[`<LinkWithIcon /> renders correctly 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 span:first-child {
  display: inline-block;
  width: 17px;
  margin-right: 12px;
}

.emotion-0 svg {
  fill: currentColor;
  display: block;
  margin: 0 auto;
}

<Link
  className="dh-link--icon emotion-0"
  external={false}
  noFollow={false}
  to="https://deringhall.com"
>
  <span>
    <i>
      custom-icon
    </i>
  </span>
  <span>
    Label
  </span>
</Link>
`;
